app/lib/actions/candlepin/environment/set_content.rb

```diff
@@ -19,18 +19,32 @@ class SetContent < Candlepin::Abstract
           param :content_ids, Array
         end
 
-        def run
-          saved_cp_ids = ::Katello::Resources::Candlepin::Environment.
+        def existing_ids
+          ::Katello::Resources::Candlepin::Environment.
               find(input[:cp_environment_id])[:environmentContent].map do |content|
             content[:contentId]
           end
+        end
+
+        def run
+          saved_cp_ids = existing_ids
           add_ids    = input[:content_ids] - saved_cp_ids
           delete_ids = saved_cp_ids - input[:content_ids]
-
-          output[:add_ids]         =    add_ids
-          unless add_ids.empty?
-            output[:add_response]    = ::Katello::Resources::Candlepin::Environment.
-              add_content(input[:cp_environment_id], add_ids)
+          retries = 2
+          (retries + 1).times do
```

Comment:
We might use the `retry` ruby keyword

```ruby
retries = 2
begin
  add_ids = input[:content_ids] - existing_ids
  unless add_ids.empty?
    output[:add_response] = ::Katello::Resources::Candlepin::Environment.add_content(input[:cp_environment_id], add_ids)
  end
rescue RestClient::InternalServerError
  retries -= 1
  retry if retries >= 0
end
```

And then extract the retry into a method:

```ruby
def with_retry(retries = 2)
  begin
    yield
  rescue RestClient::InternalServerError
    retries -= 1
    retry if retries >= 0
  end
end
```
so that the code simplifies to this:

```ruby
with_retries do
  add_ids = input[:content_ids] - existing_ids
  unless add_ids.empty?
    output[:add_response] = ::Katello::Resources::Candlepin::Environment.add_content(input[:cp_environment_id], add_ids)
  end
end
```